### DISCUSSION TITLE
What is a lint rule made of?

### DISCUSSION THREAD

**Discussion by a-frantz** [stjude-rust-labs/wdl#12] [Rule Proposals]
Below are my thoughts on the necessary components of a lint rule. Ideally, whoever proposes a new rule would fully define each of the below pieces. Together, they can be implemented into a lint rule.
* The **title** is a name for the rule, in snake_case
* The **subject** is a one sentence summary of the rule
* The **body** is a full-text description of the rule. It contains all the details needed to understand, comply with, and enforce the rule.
* The **fix** is _optional_ and is a helpful message to the user about how to comply with the rule
* At least one **tag** is required and groups similar rules together. The purpose of a tag is to organize documentation and make it easy for users to toggle categories of related rules on/off. There is no limit to how many tags may be associated with a rule.
  * examples: spacing, naming, sorting, completeness, container, and deprecated
* The **level** is one of Low/Medium/High.
  * **Low** means the rule is arbitrary or a matter of style
  * **High** means what the user has written may have unintended consequences
  * **Medium** is everything else. This is commonly issues of clarity, portability, etc.
* The **examples** are _optional_ and are either "Good" or "Bad"

Ideally, everything listed here would make it into Rust; either as literal code or as part of doc strings. We should aim for having documentation which is parsed from the code. The code will be a dependable source of truth.

**Comment by a-frantz** [comment]
@peterhuene @adthrasher I want to discuss the implementations for `subject`, `body`, and `fix`. The most straightforward implementation is that each `Rule` will have static text for these 3. I don't see any reason to complicate it further. There will be different conditions under which those 3 text fields are shown to the user. Those different conditions will inform how exactly we should define these fields.

Above, I say subject should be one sentence, but I think that slightly misses the mark because it leaves open the potential for run-on sentences that defeat the purpose of the field. We want `subject` to describe the issue **in as few characters as possible** so it can be easily displayed on one terminal line or in a small text box within an IDE. When asking rule or code contributors to write a subject I think the language we want to use is `a brief sentence fragment describing the core issue`. Should there be a hard character limit?

`body` I think would be more-or-less hidden from users under most conditions. I envision a `sprocket explain <lint code or lint name>` command which will write out the `body` of a lint rule. `body` would also form the bulk of the text in downstream documents/webpages/wikis/etc. The length of a body is going to almost entirely depend on the complexity of a given rule. I'm not sure we can provide much more guidance on length, other than what I say above (`It contains all the details needed to understand, comply with, and enforce the rule.`)

I'm not entirely sure how to define `fix`. One sentence should suffice in most cases, right? Is there a length that is too long? Does _every_ rule need a fix?
~For now, I think `fix` should always be a complete sentence (as opposed to `subject` which can be a sentence fragment to reduce character count) and every rule should have one. We can reassess if this proves difficult for any future rule. IDK, I don't feel confident on this one. Open to other thoughts.~
**EDIT**: I've realized that how we are currently using `fix` in the code is better than what I propose here. `fix` is best when it's context-dependent. There may be more than one fix for any given rule. The fix communicated to the user will depend on the user's code. See the code for the `snake_case` rule for a good example.
My gut says fix details can be ignored during the discussion phase of a rule proposal, and only needs to be addressed during the code implementation. But should it be required in all cases where a lint is thrown? I think for now we require it, and we can always revisit that requirement.

When we come to an agreement on this, I'll start a PR with whatever changes we decide. I will also update the PR/discussion templates and any other documentation to reflect the decisions reached here.

Note, I have intentionally left out the discussion of `examples`. I think those will be much more complicated to turn into Rust code than these 3 aspects of lint rules. So I'm just kicking that can down the road 🤷

**Comment by peterhuene** [comment]
Adding my thoughts to the above as I'm now consolidating the lint rules and migrating them to the new AST representation.

I think there should be separation between a _rule_ and the _diagnostics_ a rule can output, as it shouldn't need to be 1:1.

Here's what I think is relevant to a _rule_:

* A snake case `id` or `name` that uniquely identifies that rule for a particular version of WDL; a rule's id will show up in linting configuration and possibly comments in the source to disable the rule.
* A `description` or `subject` of the rule that is a single sentence describing the rule's purpose; this might be displayed in a list of available rules, for example.
* An `explanation` or `body` that is the long-form in-depth description of the rule and why it exists; this might be displayed in a command like `sprocket explain`.
* A set of `tags` that categorize the lint rule; these can be used to filter the rules to run and displayed in commands like `sprocket explain`.
* An optional `url` to link to more information about the rule, including examples and fixes; the `url` can also be printed if the rule emits any diagnostics, allowing users to easily click the lint in a terminal for more information.

I think we should create an [mdbook](https://github.com/rust-lang/mdBook) for the rules, put it up on GitHub pages, and link to them from our rules via `url` too.

Here's what I think is relevant to a rule _diagnostic_, modeled off of other `Diagnostic` representations in the wild:

* A snake case `rule` to correlate the diagnostic with the rule that is emitting it.
* A `severity` level, which is typically `error`, `warning`, or `note`; I think that this better encapsulates the existing `high`, `medium`, and `low` levels in that: errors are definite problems (a missing `runtime` section, for example), warnings are potential problems, and notes are more style than substance; this also allows for the diagnostics to colorize differently and not treat all diagnostics as warnings like we currently do.
* A `message` that is a single sentence describing the diagnostic.
* An optional `fix` describing the way to fix the problem as a single sentence.
* An optional set of labeled spans with exactly one in the set marked as "primary"; these will allow the diagnostic to annotate the source with messages; the primary span and will be used as the location of the diagnostic when printing a "short form" message.

Currently our diagnostic formatting is fairly verbose:

```
warning: <version>::<code>::[<tags>]::<level>: <message>
<optional-labeled-spans> (we currently just repeat the same message for all spans)
<optional-fix>
<see-also-sprocket-explain>
```

I think we can remove `version` entirely as it's not relevant to users and the with the new parser, one doesn't have to specify the version of the file you're linting, it just automatically runs the appropriate lint rule set based on the version of the file that was parsed.

The `code` is numerical and not particularly helpful as we already have the id of the lint and (hopefully) a url users can click on for more information; there's a reason that clippy doesn't emit numerical codes as they're not easy to search and inherently have no meaning when you see one. Additionally, we have to maintain in the source that they're unique, which is burdensome as the list grows.

Including the `tags`, as it is a list, really makes the message unwieldy and requires quite a bit of eye travel to get at relevant information for users (i.e. "oi, what's all this then?"). I really think we should not include this in diagnostic messages.

The `level`, to me, is synonymous with existing severity in `codespan_reporting` and redundant.

Therefore, I propose we use a long format like:

```
<severity>[<rule-id>]: <message>
<optional-labeled-spans>
<optional-fix>
<optional-see-this-url-for-more-information>
<see-also-sprocket-explain>
```

and a short format like this (which corresponds to [DisplayStyle::Short](https://docs.rs/codespan-reporting/latest/codespan_reporting/term/enum.DisplayStyle.html#variant.Short) in `codespan-reporting`:

```
<optional-primary-location>: <severity>[<rule-id>]: <message>
```

**Reply by a-frantz** [reply]
More-or-less agree with everything written here. (To be frank, I like what you present here more than what I started this thread with)

Before reading this I was struggling to come up with the vocab used here, specifically the division between "Rule" and "Diagnostic". So definitely on board with more formally separating those concepts. I think in "today's" code those concepts are a bit blurred together and that was causing me issues when rationalizing about all this.

I think the most controversial idea is transferring `Level` to [`Severity` as defined by `codespan`](https://docs.rs/codespan-reporting/latest/codespan_reporting/diagnostic/enum.Severity.html). I'm not in love with our current concept of `Level`. My complaint is that `Low`/`Medium`/`High` are not intuitively defined. I proposed functional definitions above that I'm ok with, but I would prefer to move to a system that's more intuitive for our users. However, I'm not sure that we can pull `Severity` straight off the shelf 🤔 
- `Bug` - I'm not aware of any cases where we'd use this?
- `Error` -  I _would not_ like calling any of our lints "errors". Even the ones we've labelled `High`. I want us to be quite clear in terms of what's a violation of the specification and what's our opinion of "good" WDL. None of our lints will be violations of the spec (by nature), and that makes the term `Error` inapplicable.
- `Warning` - IMO any lint we currently label `High` would become a `Warning`. Maybe even some (or all) of the `Medium` lints?
- `Note` - This seems to correspond pretty cleanly with our `Low`
    - If `Note == Low` and `Warning == High` we lose the label `Medium`. Medium lints will have to be upgraded or downgraded.
- `Help` - Would we use `Help`? Maybe for the deprecation lints?
If we use the `codespan` `Severity`, we will be practically making use of only 2 out of 5 enum variants. That's a bad fit IMO.

**Reply by adthrasher** [reply]
> * `Error` -  I _would not_ like calling any of our lints "errors". Even the ones we've labelled `High`. I want us to be quite clear in terms of what's a violation of the specification and what's our opinion of "good" WDL. None of our lints will be violations of the spec (by nature), and that makes the term `Error` inapplicable.

This was also my immediate thought. Otherwise, I think what Peter outlines makes sense.

**Reply by peterhuene** [reply]
Thinking about this more, perhaps `severity` doesn't belong on the diagnostic at all?

Clippy, like `rustc`, has the concept of `allow`, `warn`, and `deny`, which is configured per-rule. If `allow` the rule is disabled, if `warn` it emits a warning, and if `deny` it emits an error.

The beneficial thing about this design is that it is _configuration_ that determines if we treat a lint rule violation as something that would cause a non-zero exit from the linter (should be non-zero on _errors only_, something we don't do today); thus users get to decide if they never want to see a single quoted string, for example, in their source base by denying them.

Perhaps a rule should specify its default configuration state of `allow`, `warn`, and `deny` (whatever we call this).

If we want `warn` to emit either warnings or notes, then perhaps each diagnostic can have a `priority` on it, where `Priority::Normal` => warning and `Priority::Low` => note?

**Reply by a-frantz** [reply]
Interesting ideas here.

> perhaps severity doesn't belong on the diagnostic at all?

I think this is correct. It belongs to the `Rule` not the `Diagnostic`. Being emitted as part of the diagnostic sounds about as useful as emitting tags.

> Perhaps a rule should specify its default configuration state of allow, warn, and deny (whatever we call this).

I'm on board with this approach to configuration. But I think it's separate from `Severity`/`Level`.

The point of `Level` is to give users a way to specify something like `--only-high` and then only lints we think are _critical_ in some way (such as missing a `runtime` or mixing spaces and tabs within a `command`) will get reported. It's very similar in purpose to tags, but just another axis for slicing and dicing the lint rules. Maybe we don't need 3 labels, we could just have a `critical` boolean for each Rule? That strikes me as more intuitive at least.

**Reply by peterhuene** [reply]
> The point of Level is to give users a way to specify something like --only-high and then only lints we think are critical in some way (such as missing a runtime or mixing spaces and tabs within a command) will get reported. It's very similar in purpose to tags, but just another axis for slicing and dicing the lint rules. Maybe we don't need 3 labels, we could just have a critical boolean for each Rule? That strikes me as more intuitive at least.

Could we represent this with a `Tag::Important` or some such to indicate the rule is considered important enough by the tool to run when you specify `--only-important` (effectively implies `--tags important`)?

On the surface, tags feel like the appropriate place to filter to me without needing further abstraction.

**Reply by a-frantz** [reply]
That might be the right answer there. I'd like to chew on all this before we set anything in stone. But `Tag::Important` feels like the right direction.

**Reply by a-frantz** [reply]
> Clippy, like `rustc`, has the concept of `allow`, `warn`, and `deny`, which is configured per-rule. If `allow` the rule is disabled, if `warn` it emits a warning, and if `deny` it emits an error.
> 
> The beneficial thing about this design is that it is _configuration_ that determines if we treat a lint rule violation as something that would cause a non-zero exit from the linter (should be non-zero on _errors only_, something we don't do today); thus users get to decide if they never want to see a single quoted string, for example, in their source base by denying them.
> 
> Perhaps a rule should specify its default configuration state of `allow`, `warn`, and `deny` (whatever we call this).

Looking at this again, I initially said I'd be on board with this direction but I'm not sure anymore. This seems like it might be _too much_ configuration. I don't see the practical distinction between `warn` and `deny`. How I approach my linting, I consider any output from the linter an issue that _must be addressed_ either via a code change or by silencing that lint. So I don't see any value to having _both_ `warn` and `deny`.

I only see the need for two states for a lint, "run" and "don't run": _any rule violation leads to a non-zero exit code_ (although maybe we have a `--tolerant` flag that only uses non-zero exit code for "real" (non-lint) errors?).

Sidebar: I think we probably want to add a new subcommand to sprocket, `check` or `validate`, that only performs WDL-specification validation and ignores all lints.

**Reply by adthrasher** [reply]
@a-frantz - If we have a `check`/`validate` subcommand, I think the argument for lint warnings in the `lint` subcommand to exit non-zero is much stronger. That also potentially removes the need for something like `--tolerant` in the `lint` subcommand. If you want to ignore lint warnings, you could just run the non-lint version.

**Reply by adthrasher** [reply]
I dislike the idea of `Tag::Important`. It feels a bit like overloading the `tag` functionality. I'm not certain we need the severity concept at all. How do we envision someone using it? I doubt that a user will come along and disagree with everything in our `Low` bucket and want to disable them all. Even for controversial things (`tabs` vs `spaces` or single vs double quotes), I doubt that you could create a set that would always be appropriate. I imagine that there will be a user for every combination of those.

I think I'd be in favor of having minimal grouping methods (just `tags`?) to start. If we later determined that additional axes for groupings would be useful, it would be straightforward to add.

**Reply by a-frantz** [reply]
@adthrasher 
>  I doubt that a user will come along and disagree with everything in our Low bucket and want to disable them all.

This seems true and is a good argument against the existence of the `Low` category. But what about my `--only-high`/`--only-important`/`--only-critical` example? That seems like a real use case to me.

> I think I'd be in favor of having minimal grouping methods (just tags?) to start. If we later determined that additional axes for groupings would be useful, it would be straightforward to add.

I think this right here is an argument _for_ the "overloading" of tags with the concept of `important` (or whatever we decide to call the rules we are currently labelling `High`. `important` seems wrong to me, but I'm not sure what's better)

**Reply by adthrasher** [reply]
I think the `high`/`important`/`critical` is just a different version of the same thing. If we don't expect users to bulk disable `Low`, would we expect them to bulk disable `Medium` and `Low`? 

That said, is this something we would even want to encourage? In an ideal world, the lint rules decided here would come to dominate, so everyone would use them. I'd prefer not to encourage people getting in the habit of disabling a full set of rules. If `--fix` is implemented, I'd prefer the experience of disabling a large number of rules to be unpleasant to encourage conformance.

**Reply by a-frantz** [reply]
> That said, is this something we would even want to encourage? In an ideal world, the lint rules decided here would come to dominate, so everyone would use them. I'd prefer not to encourage people getting in the habit of disabling a full set of rules. If --fix is implemented, I'd prefer the experience of disabling a large number of rules to be unpleasant to encourage conformance.

This is thought provoking. I think I agree with all this, but doesn't it suggest we shouldn't have tags either? I think the exception for tags is that they're useful for organizing documentation. Can we have tags in our docs but not exposed on the CL? Is that antagonizing users? 😂

**Reply by adthrasher** [reply]
That is a good point. I am less opposed to exposing tags as a filter because the tags themselves are useful for documentation purposes. So we will have tags either way. Maybe we should omit the CLI functionality though. It would be relatively easy to add in later if we had a compelling reason since the tags will exist.

**Reply by peterhuene** [reply]
> Looking at this again, I initially said I'd be on board with this direction but I'm not sure anymore. This seems like it might be too much configuration. I don't see the practical distinction between warn and deny. How I approach my linting, I consider any output from the linter an issue that must be addressed either via a code change or by silencing that lint. So I don't see any value to having both warn and deny

One thing the Rust design affords is that new versions of the tool, with new lints coming online, do not suddenly turn your CI red because of lint violations on previously accepted code; they would print as warnings and exit zero. Folks would start to see the warnings and fix them or decide those warnings really should be errors and deny them for the future.

I don't know if that's something to consider for us, though; Rust is a general purpose programming language while `wdl` is relatively domain-specific. All I do know is that I really appreciate the efforts the Rust team puts into things for making a `rustup update` rarely a painful experience.

**Reply by a-frantz** [reply]
Ahh. That's a consideration I hadn't thought of. Great point Peter.

Potentially crazy idea: Can we put new rule's in some sort of "beta" phase? Where they "warn" without causing a non-zero exit code? The first release a rule is "live" it's considered in this "beta phase", and then subsequent release would upgrade to regular status?

I guess that only works if we expect everyone to stay on the bleeding edge. That's not a great solution for people who wait several releases between updates... 🤔 

(That would be annoying to manage with code, unless we tracked which release each was added and then compared that to the current version.)

**Reply by adthrasher** [reply]
@peterhuene - I may have missed this earlier in the thread, but how does Rust avoid breaking CI with new lints?

**Reply by a-frantz** [reply]
Another benefit to my crazy idea above: The current process we've (loosely) imagined for new lints is starting with a discussion thread -> upgrading to an issue -> PR gets merged with the new lint -> new release it's "live". We want there to be some form of feedback stage before implementation. But this also assumes a somewhat active community of WDL authors paying attention to our repo _before any code is written_. We can aspire to that, but we aren't there today.

A more realistic approach would be to have a feedback stage during this "beta phase" I suggested (shopping for new names if we're taking this idea seriously). We might roll something out, a bunch of people's CI is printing new warnings (but not turning RED) and we receive backlash. In that case we'd reconsider upgrading the rule to `deny`.

**Reply by peterhuene** [reply]
@adthrasher

If `clippy` introduces a new default `warn` lint, upgrading your Rust toolset in CI might cause it to print a warning if your existing source violates that new lint, but the tool will still exit zero.

Only if you denied _all_ lints by configuration would upgrading your Rust toolset break your linting CI job.

It's pretty flexible in how you use the tool: you can treat linting as prescriptive (deny all), opt-in to things you never want see in your code base (deny some), or treat it merely as advisory (deny none), all while allowing you to completely ignore specific rules.

**Reply by adthrasher** [reply]
Ah, so all new lints are `warn`. Then it's up to the user for every single lint to opt-in or opt-out.

**Reply by peterhuene** [reply]
It depends on the lint; clippy does have some default `deny` lints, but those are usually reserved for "there's no reason to ever do what you're doing".

**Reply by peterhuene** [reply]
Indeed, the default deny lints are from the `correctness` category (like doing a `let pi = 3.14`).

**Reply by a-frantz** [reply]
I'm changing directions. Hot take inbound: 🔥 

If we have `sprocket validate` *and* `sprocket lint`, I would like the default behavior of `lint` to be "yell at the user when any lint is violated". If you don't want to be yelled at, try `sprocket validate` or `sprocket lint --tolerant` (so your CI stays green but prints every warning).

Isn't our ultimate goal to *get people adopting our lints*? Why are we writing these if we expect people to ignore them? Let's yell at people 😂 Breaking their CI when they violate our rules *is a good thing*. We are trying to drive adoption! Let's be bold.

**Reply by peterhuene** [reply]
I'm onboard with that approach :+1:

That said, I do like `check` ("check that your source is valid") over `validate` ("validate your source"); I don't think I can articulate a reason why, though.

One thing to think about though is whether or not a check command would ever warn; currently all the diagnostics emitted by parsing and validating are errors and should non-zero exit, but what if it warned for things like an unused declaration, which historically is a diagnostic more in the realm of the compiler than the linter.

**Reply by a-frantz** [reply]
I'm leaning toward `validate` because there already exists `miniwdl check` which does the same thing. I want to differentiate us from miniwdl, but IDK maybe using the same vocab as miniwdl is a good thing?

EDIT: similarly I've thought instead of (or in addition to) `sprocket lint --fix` we could have `sprocket fmt`, in the same vein as `cargo fmt`. IDK if matching existing tools is good or bad, but I _feel_ like it's probably bad? No reasoning for that.

**Reply by adthrasher** [reply]
> I'm leaning toward `validate` because there already exists `miniwdl check` which does the same thing. I want to differentiate us from miniwdl, but IDK maybe using the same vocab as miniwdl is a good thing?

FWIW, `womtool` uses `validate`. So either way we're overlapping one of the two main players.

**Reply by a-frantz** [reply]
> One thing to think about though is whether or not a check command would ever warn

By default, I think every diagnostic emitted by check/validate/whatever would be an error and result in a non-zero exit code. Any violation of the spec should be considered as meaning "your code is **broken** and _can't be run_." Is there ever a situation where that meaning should correspond to a zero exit code?

> but what if it warned for things like an unused declaration, which historically is a diagnostic more in the realm of the compiler than the linter.

Hmmm... That's an interesting question. IDK if I can articulate _why_, but in my mind `check`/`validate` would only be concerned with violations of the WDL specification. But should `lint` cover unused declarations??? Is that OOS for a linter? I don't think it goes anywhere else, and we don't want to introduce another command just for these compiler-esque-warnings, so I think they have to be folded into `lint`... Not a perfect fit, but I think it's ok?

--------------
> FWIW, womtool uses validate. So either way we're overlapping one of the two main players.

wompwomp
